### PR TITLE
CFE-1050: Add support of capacity reservation group in openshift/machine-api-provider-azure

### PR DESCRIPTION
**Pull request by anirudhAgniRedhat**
Add support of capacity reservation group in openshift/machine-api-provider-azure so that azure VMs can be associated to a capacity reservation group during the VM creation.

Implementation PR for https://github.com/openshift/api/pull/1866

### CODE REVIEW

**Comment by anirudhAgniRedhat** [comment]
I have to update the go.mod file to replace openshift/api dependency as the api PR is not merged yet. 
I will update the vendor and go.mod file once the reviewers are good with the implementations.
we can merge API PR and remove dependency.

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 1]
```diff
@@ -681,6 +681,13 @@ func (s *Reconciler) createVirtualMachine(ctx context.Context, nicName, asName s
 			vmSpec.ManagedIdentity = azure.GenerateManagedIdentityName(s.scope.SubscriptionID, s.scope.MachineConfig.ResourceGroup, s.scope.MachineConfig.ManagedIdentity)
 		}
 
+		if s.scope.MachineConfig.CapacityReservationGroupID != "" {
+			if err = validateCapacityReservationGroupID(s.scope.MachineConfig.CapacityReservationGroupID); err != nil {
+				return fmt.Errorf("invalid capacityReservationGroupID : %w", err)
```
Comment (TrilokGeer) [inline pkg/cloud/azure/actuators/machine/reconciler.go:686]: the error message should be describe as a failure event. e.g: "failed to validate capacity reservation group id : %w"

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 2]
```diff
@@ -47,7 +47,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/klog/v2"
 	"k8s.io/utils/ptr"
-
+	azureutils "sigs.k8s.io/cluster-api-provider-azure/util/azure"
```
Comment (JoelSpeed) [inline pkg/cloud/azure/actuators/machine/reconciler.go:50]: I would rather not introduce this new dependency, can you copy the util over instead please?

**Review by JoelSpeed — COMMENTED** [review]
Azure webhook validation lives in a different repository, please see [here](https://github.com/openshift/machine-api-operator/blob/ac84cecd1374041a1dbb70313f76504c41b131a6/pkg/webhooks/machine_webhook.go#L861-L962) and raise a PR for the relevant validation there as well

**Comment by anirudhAgniRedhat** [comment]
> Azure webhook validation lives in a different repository, please see [here](https://github.com/openshift/machine-api-operator/blob/ac84cecd1374041a1dbb70313f76504c41b131a6/pkg/webhooks/machine_webhook.go#L861-L962) and raise a PR for the relevant validation there as well

@JoelSpeed You mean I should remove the validation for parseResourseID from here and create a seperate PR for adding this Validation in that PR right?

Also write the function itself for validation to remove its dependency.

**Comment by anirudhAgniRedhat** [comment]
> It would be preferable to have both controller and webhook side validation where possible, can you copy over from the webhook back to here?

@JoelSpeed I have added back the validation here. Regarding unit test fail it is again due to same reason as I pointed in https://github.com/openshift/machine-api-operator/pull/1234#issuecomment-2088859055

Here The suite test is not able to create machineset 
```
[BeforeSuite] 
/home/aagnihot/workspace/downstream/openshift-machine-api-provider-azure/pkg/cloud/azure/actuators/machineset/controller_suite_test.go:57
  [FAILED] in [BeforeSuite] - /home/aagnihot/workspace/downstream/openshift-machine-api-provider-azure/pkg/cloud/azure/actuators/machineset/controller_suite_test.go:87 @ 05/02/24 12:32:55.143
[BeforeSuite] [FAILED] [2.739 seconds]
[BeforeSuite] 
/home/aagnihot/workspace/downstream/openshift-machine-api-provider-azure/pkg/cloud/azure/actuators/machineset/controller_suite_test.go:57

  [FAILED] Expected success, but got an error:
      <*meta.NoKindMatchError | 0xc0007b0200>: 
      no matches for kind "Infrastructure" in version "config.openshift.io/v1"
      {
          GroupKind: {
              Group: "config.openshift.io",
              Kind: "Infrastructure",
          },
          SearchedVersions: ["v1"],
      }
  In [BeforeSuite] at: /home/aagnihot/workspace/downstream/openshift-machine-api-provider-azure/pkg/cloud/azure/actuators/machineset/controller_suite_test.go:87 @ 05/02/24 12:32:55.143

  Full Stack Trace
    github.com/openshift/machine-api-provider-azure/pkg/cloud/azure/actuators/machineset.init.func1()
        /home/aagnihot/workspace/downstream/openshift-machine-api-provider-azure/pkg/cloud/azure/actuators/machineset/controller_suite_test.go:87 +0xb95
------------------------------
[AfterSuite] 
/home/aagnihot/workspace/downstream/openshift-machine-api-provider-azure/pkg/cloud/azure/actuators/machineset/controller_suite_test.go:103
[AfterSuite] PASSED [1.081 seconds]
------------------------------

Summarizing 1 Failure:
  [FAIL] [BeforeSuite] 
  /home/aagnihot/workspace/downstream/openshift-machine-api-provider-azure/pkg/cloud/azure/actuators/machineset/controller_suite_test.go:87

Ran 0 of 8 Specs in 3.822 seconds
FAIL! -- A BeforeSuite node failed so all tests were skipped.
--- FAIL: TestReconciler (3.83s)
FAIL
```

File: pkg/cloud/azure/actuators/machineset/controller_suite_test.go [hunk 3]
```diff
@@ -25,7 +25,9 @@ import (
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 	configv1 "github.com/openshift/api/config/v1"
+	_ "github.com/openshift/api/config/v1/zz_generated.crd-manifests"
 	machinev1 "github.com/openshift/api/machine/v1beta1"
+	_ "github.com/openshift/api/machine/v1beta1/zz_generated.crd-manifests"
```
Comment (JoelSpeed) [inline pkg/cloud/azure/actuators/machineset/controller_suite_test.go:30]: Should be in the tools.go

**Comment by JoelSpeed** [comment]
I think this is now ready to be tested with the other 2 PRs by QE

**Pull request merged**